ako-infra/proxy/management_service_controller.go

```diff
@@ -0,0 +1,703 @@
+	stopCh    chan struct{}
+}
+
+var namespaceProcessor *NamespaceGrantProcessor
+
+func StartNamespaceGrantProcessor() {
+	if namespaceProcessor == nil {
+		workqueue := workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(), //nolint:staticcheck
+			VKSNamespaceWorkQueue,
+		)
+
+		namespaceProcessor = &NamespaceGrantProcessor{
+			workqueue: workqueue,
+			stopCh:    make(chan struct{}),
+		}
+
+		go namespaceProcessor.runWorker()
+		utils.AviLog.Infof("VKS ManagementServiceGrant: namespace processor initialized with retry queue")
+
+		// Perform one-time reconciliation at startup to ensure all grants exist
+		go func() {
+			time.Sleep(5 * time.Second)
+			ReconcileManagementServiceGrants()
+		}()
+	}
+}
+
+func StopNamespaceGrantProcessor() {
+	if namespaceProcessor != nil {
+		close(namespaceProcessor.stopCh)
+		namespaceProcessor.workqueue.ShutDown()
+		namespaceProcessor = nil
+		utils.AviLog.Infof("VKS ManagementServiceGrant: namespace processor stopped")
+	}
+}
+
+func (p *NamespaceGrantProcessor) runWorker() {
+	for p.processNextWorkItem() {
+	}
+}
+
+func (p *NamespaceGrantProcessor) processNextWorkItem() bool {
+	obj, shutdown := p.workqueue.Get()
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer p.workqueue.Done(obj)
+
+		event, ok := obj.(*NamespaceEvent)
+		if !ok {
+			p.workqueue.Forget(obj)
+			utils.AviLog.Errorf("VKS ManagementServiceGrant: expected NamespaceEvent in workqueue but got %#v", obj)
+			return nil
+		}
+
+		if err := p.processNamespaceEvent(event); err != nil {
+			p.workqueue.AddRateLimited(event)
+			return fmt.Errorf("error processing namespace event %s for %s: %s, requeuing", event.EventType, event.Namespace, err.Error())
+		}
+
+		p.workqueue.Forget(obj)
+		utils.AviLog.Infof("VKS ManagementServiceGrant: successfully processed %s event for namespace %s", event.EventType, event.Namespace)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utils.AviLog.Errorf("VKS ManagementServiceGrant: %v", err)
+		return true
+	}
+
+	return true
+}
+
+// processNamespaceEvent handles the actual namespace event processing
+func (p *NamespaceGrantProcessor) processNamespaceEvent(event *NamespaceEvent) error {
+	if !lib.GetVPCMode() || !lib.IsVKSCapabilityActivated() {
+		return nil
+	}
+
+	controller := NewManagementServiceController()
+	if controller == nil {
+		return fmt.Errorf("failed to create ManagementServiceController")
+	}
+
+	switch event.EventType {
+	case NamespaceEventAdd:
+		if event.NewHasSEG {
+			if err := controller.CreateManagementServiceGrant(event.Namespace); err != nil {
+				return err
+			}
+			utils.AviLog.Infof("VKS ManagementServiceGrant: successfully created grant for namespace %s", event.Namespace)
+			return nil
+		}
+	case NamespaceEventUpdate:
+		if !event.OldHasSEG && event.NewHasSEG {
+			// SEG annotation was added
+			if err := controller.CreateManagementServiceGrant(event.Namespace); err != nil {
+				return err
```

Comment:
How are we handling the error here as Reconciliation is run only in boot up flow?